src/object.jl

```diff
@@ -372,3 +373,121 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
     end
     return b
 end
+
+function _upload_task(io, lk; kw...)
+    try
+        upload_buffer = take!(io.upload_queue)
+        # upload the part
+        parteTag, wb = uploadPart(io.store, io.url, upload_buffer, io.cur_part_id, io.uploadState; io.credentials, kw...)
+        # add part eTag to our collection of eTags in the right order
+        lock(lk) do
+            put!(io.sync, io.cur_part_id) do
+                push!(io.eTags, parteTag)
+            end
+        end
+        # atomically increment our part counter
+        @atomic io.cur_part_id += 1
```

Comment:
Indeed I've seen this happening. On the other hand if we increment before `put!` two tasks might have incremented `io_cur_part_id` before reaching `put!` and `io.sync.i`, which is incremented inside `put!` will be != io.cur_part_id` resulting in a deadlock.

Reply:
I fixed this by assigning `io.cur_part_id`to a variable and passing it as a parameter to `_upload_task` rather than getting its value from `io.cur_part_id` directly.

Reply:
Yes, making this the callers problem seems like a nice solution to me :+1: